hoomd/comm.py

```diff
@@ -379,3 +380,29 @@ def barrier(self):
         if _hoomd.is_MPI_available():
             self.cpp_mpi_conf.barrier()
 
+    @contextlib.contextmanager
+    def localize_abort(self):
+        """ Localize MPI_Abort to this partition.
+
+        HOOMD calls MPI_Abort to tear down all running MPI processes whenever
+        there is an uncaught exception. By default, this will abort the entire
+        MPI execution. When using partitions (``nrank is not None``), an uncaught
+        exception on one partition will therefore abort all of them.
+
+        Use the return value of :py:meth:`localize_abort()` as a context manager
+        to tell HOOMD that all operations within the context will use only
+        that MPI communicator so that an uncaught exception in one partition
+        will only abort that partition and leave the others running.
+        """
+
+        global _current_communicator
+        prev = _current_communicator
+
+        _current_communicator = self
+        yield None
```

Comment:
Given the other conversation below about context managers, should this be wrapped in a `try/finally`?

Reply:
I tried, but it failed tests. By the time the uncaught exception gets to the exception hook, the `finally:` block has executed and reset the previous communicator - which is exactly what we don't want to happen. In this case we actually want the `_current_communicator` to remain set and there is no need to reset it while the process is being torn down.